src/main/java/com/mixpanel/android/mpmetrics/MixpanelPushNotification.java

```diff
@@ -301,28 +254,99 @@ protected void maybeAddActionButtons() {
+                    JSONObject buttonObj = buttonsArr.getJSONObject(i);
+
+                    // get button icon from name if one sent
+                    int btnIcon = NotificationData.NOT_SET;
+                    if (buttonObj.has("icnm")) {
+                        String btnIconName = buttonObj.getString("icnm");
+                        if (drawableIds.knownIdName(btnIconName)) {
+                            btnIcon = drawableIds.idFromName(btnIconName);
+                        }
+                    }
+
+                    // handle button label
+                    final String btnLabel = buttonObj.getString("lbl");
+
+                    // handle button action
+                    final JSONObject pushActionJSON = buttonObj.getJSONObject("ontap");
+                    final PushTapAction pushAction = new PushTapAction(pushActionJSON.getString("type"), pushActionJSON.getString("uri"));
+
+                    //handle button id
+                    final String btnId = buttonObj.getString("id");
+
+                    buttons.add(new NotificationButtonData(btnIcon, btnLabel, pushAction, btnId));
+                }
+            } catch (JSONException e) {
+                MPLog.e(LOGTAG, "Exception parsing buttons payload", e);
+            }
+        }
+    }
+
+    protected PushTapAction buildOnTap(String onTapStr, String uriString) {
+        PushTapAction onTap;
+        if (null != onTapStr) {
+            try {
+                final JSONObject onTapJSON = new JSONObject(onTapStr);
+                onTap = new PushTapAction(onTapJSON.getString("type"), onTapJSON.getString("uri"));
+            } catch (JSONException e){
+                MPLog.e(LOGTAG, "Couldn't parse JSON Object for \'mp_ontap\'");
+                onTap = null;
+            }
+        } else {
+            onTap = null;
+        }
+
+        if (null == onTap) {
+            if (null == uriString) {
+                onTap = new PushTapAction("homescreen", "");
+            } else {
+                onTap = new PushTapAction("browser", uriString);
+            }
+        }
+
+        return onTap;
+    }
+
     @TargetApi(20)
     protected Notification.Action createAction(int icon, CharSequence title, PushTapAction onTap, String actionId) {
-        return (new Notification.Action.Builder(icon, title, createActionIntent(onTap, actionId))).build();
+        return (new Notification.Action.Builder(icon, title, createActionIntent(onTap, actionId, title))).build();
     }
 
-    protected PendingIntent createActionIntent(PushTapAction onTap, String actionId) {
+    protected PendingIntent createActionIntent(PushTapAction onTap, String actionId, CharSequence label) {
+        Intent routingIntent = getRoutingIntent(onTap, actionId, label);
+        return PendingIntent.getActivity(context, 0, routingIntent, PendingIntent.FLAG_CANCEL_CURRENT);
+    }
+
+    protected Intent getRoutingIntent(PushTapAction onTap, String actionId, CharSequence label) {
```

Comment:
yeah let's make this signature: `protected Intent getRoutingIntent(PushTapTarget target, PushTapAction onTap, String buttonId = null, CharSequence label = null)`

Reply:
if target is already a member of onTap then what is the point in putting it in the signature?